openmmapi/src/NoseHooverIntegrator.cpp

```diff
@@ -0,0 +1,354 @@
+    nhcKernel = context->getPlatform().createKernel(NoseHooverChainKernel::Name(), contextRef);
+    nhcKernel.getAs<NoseHooverChainKernel>().initialize();
+    forcesAreValid = false;
+
+    // check for drude particles and build the Nose-Hoover Chains
+    for (auto& thermostat: noseHooverChains){
+        // if there are no thermostated particles or pairs in the lists this is a regular thermostat for the whole (non-Drude) system
+        if ( (thermostat.getThermostatedAtoms().size() == 0) && (thermostat.getThermostatedPairs().size() == 0) ){
+            std::vector<int> thermostatedParticles;
+            for(int particle = 0; particle < system.getNumParticles(); ++particle) {
+                double mass = system.getParticleMass(particle);
+                if ( (mass > 0) && (mass < 0.8) ){
+                    std::cout << "Warning: Found particles with mass between 0.0 and 0.8 dalton. Did you mean to make a DrudeNoseHooverIntegrator instead? "
+                                 "The thermostat you are about to use will not treat these particles as Drude particles!" << std::endl;
+                }
+                if(system.getParticleMass(particle) > 0) {
+                   thermostatedParticles.push_back(particle);
+                }
+            }
+            thermostat.setThermostatedAtoms(thermostatedParticles);
+        }
+    }
+
+    initializeThermostats(system);
+}
+
+void NoseHooverIntegrator::cleanup() {
+    vvKernel = Kernel();
+    nhcKernel = Kernel();
+}
+
+vector<string> NoseHooverIntegrator::getKernelNames() {
+    std::vector<std::string> names;
+    names.push_back(NoseHooverChainKernel::Name());
+    names.push_back(IntegrateVelocityVerletStepKernel::Name());
+    return names;
+}
+
+void NoseHooverIntegrator::step(int steps) {
+    if (context == NULL)
+        throw OpenMMException("This Integrator is not bound to a context!");
+    std::pair<double, double> scale, kineticEnergy;
+    for (int i = 0; i < steps; ++i) {
+        context->updateContextState();
+        for(auto &nhc : noseHooverChains) {
+            kineticEnergy = nhcKernel.getAs<NoseHooverChainKernel>().computeMaskedKineticEnergy(*context, nhc, false);
+            scale = nhcKernel.getAs<NoseHooverChainKernel>().propagateChain(*context, nhc, kineticEnergy, getStepSize());
+            nhcKernel.getAs<NoseHooverChainKernel>().scaleVelocities(*context, nhc, scale);
+        }
+        vvKernel.getAs<IntegrateVelocityVerletStepKernel>().execute(*context, *this, forcesAreValid);
+        for(auto &nhc : noseHooverChains) {
+            kineticEnergy = nhcKernel.getAs<NoseHooverChainKernel>().computeMaskedKineticEnergy(*context, nhc, false);
+            scale = nhcKernel.getAs<NoseHooverChainKernel>().propagateChain(*context, nhc, kineticEnergy, getStepSize());
+            nhcKernel.getAs<NoseHooverChainKernel>().scaleVelocities(*context, nhc, scale);
+        }
```

Comment:
I see it's still using the OBABO discretization rather than the BAOAB one.  Is that because you tried it and it didn't work, or have you not tried it yet?  At least with Langevin, BAOAB can allow significantly larger time steps.  I don't know whether the theory carries over to Nose-Hoover through.